### PR TITLE
Refactor DeltaLayer::dump() function

### PR DESCRIPTION
**Pull request by hlinnaka**
Put most of the code in a closure that returns Result, so that we can
use the ?-operator for error handling. That's simpler.

### CODE REVIEW

**Comment by hlinnaka** [comment]
I did this originally as part of https://github.com/neondatabase/neon/pull/1595, but it makes sense on its own so let's review and push it separately.

File: pageserver/src/layered_repository/delta_layer.rs [hunk 1]
```diff
@@ -370,34 +389,10 @@ impl Layer for DeltaLayer {
...
-                            }
-                        }
-                    }
-                    Err(err) => {
-                        write!(&mut desc, " READ ERROR: {}", err).unwrap();
-                    }
-                }
+                let desc = match dump_blob(blob_ref) {
+                    Ok(desc) => desc,
+                    Err(err) => format!("ERROR: {}", err),
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository/delta_layer.rs:394]: Is there any context around/in this `ERROR` or should we make it a bit more descriptive?

Also you can write `{err}` now.

**Review by SomeoneToIgnore — APPROVED** [review]

**Review by aome510 — APPROVED** [review]
LGTM!

File: pageserver/src/layered_repository/delta_layer.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/delta_layer.rs:394]: > Is there any context around/in this ERROR or should we make it a bit more descriptive?

I'm hoping that the error message is descriptive enough. This gets printed in lieu of the record's desription, like this:
```
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F290: rec 146 bytes, 116 compressed, will_init false: will_init: false, HEAP INSERT
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F360: rec 146 bytes, 115 compressed, will_init false: will_init: false, HEAP INSERT
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F430: rec 146 bytes, 114 compressed, will_init false: will_init: false, HEAP INSERT
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F500: ERROR: Corrupted block detected
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F5D0: ERROR: Failed to read blob from virtual file /tmp/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000001D0112069-00000001E011A4B1
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F6A0: rec 146 bytes, 114 compressed, will_init false: will_init: false, HEAP INSERT
  key 000000067F000032BE000040F10000001A57 at 1/D9B5F770: rec 146 bytes, 118 compressed, will_init false: will_init: false, HEAP INSERT
```
We're not very consistent with the style of the error messages, but for a debugging tool I think this is good enough for now.

**Pull request merged**